### ISSUE TITLE
Overloading pythons internal functions can easily lead to confusion through hanging processes

### ISSUE THREAD

**Issue by m-mohr** [Open-EO/openeo-python-client#262]
The issue originated from https://discuss.eodc.eu/t/reducing-masks-in-openeo/113.

If you pass e.g. the "sum" function to a callback and forget to import the actual process from openeo.processes you don't get an error but instead, the script hangs "forever" as it executes the internal sum function of Python (I think). I think the Python client should try to avoid this and throw an error if such an unexpected input occurs.

**Comment by Jaapel** [comment]
I know that some libraries also use `_sum` if they have functions that mirror the internal functions.

**Comment by soxofaan** [comment]
addressed in d499b377ad706d3d627ce41d46f8a09e680f7bb9:

mistakenly using builtin `sum` in a callback will now cause an exception instead of infinite loop:
```
>>> cube.reduce_dimension(dimension="t", reducer=sum)
RuntimeError: Exceeded ProcessBuilder iteration limit.
Are you mistakenly using a builtin like `sum()` or `all()` in a callback
instead of the appropriate helpers from `openeo.processes`?
```

**Comment by soxofaan** [comment]
Background info: 
apparently `sum(data)` causes the construction of an (infinite) process graph that does an infinite summation 
```python
array_element(data, 0) + array_element(data, 1) + array_element(data, 2)
 + array_element(data, 3) + ...
```

It's not really feasible to detect that builtin `sum` is used as (part of a) callback in general, but I added a check that the `array_element` index doesn't exceed a limit. The limit is 100 at the moment, which I think is generous for normal `array_element` usage. Advantage compared to trying to detect `sum` is that it will also cover other builtin cases that trigger infinite calculations like `all`, `any`, `min`, `max` (some of which currently fail with other errors).


I also added this to the docs at https://open-eo.github.io/openeo-python-client/processes.html#caveats  so that users can find background info when they google for this "Exceeded ProcessBuilder iteration limit" error
